pkg/permissionclaim/permissionclaim_labeler.go

```diff
@@ -136,3 +140,105 @@ func (l *Labeler) LabelsFor(ctx context.Context, cluster logicalcluster.Name, gr
 
 	return labels, nil
 }
+
+// ObjectPermitted determines if any PermissionClaims apply to a given object's GroupResource, namespace, or name.
+// If no APIBinding or PermissionClaim exists for the GroupResource, the object is permitted.
+func (l *Labeler) ObjectPermitted(ctx context.Context, cluster logicalcluster.Name, groupResource schema.GroupResource, resourceName, resourceNamespace string) (bool, error) {
```

Comment:
You're correct, it is. Is there a better way to do this without as much code duplication? Maybe a visitor pattern somewhere?

Reply:
I think something like this might work (probably needs some adjusting to function signatures here and there - this is just a rough sketch):

```go
func (l *Labeler) visitBindingsAndClaims(
	ctx context.Context,
	cluster logicalcluster.Name,
	groupResource schema.GroupResource,
	resourceNamespace, resourceName string,
	visit func(apiBinding *apisv1alpha1.APIBinding, claim apisv1alpha1.AcceptablePermissionClaim, namespace,
	name string),
) error {
	bindings, err := l.listAPIBindingsAcceptingClaimedGroupResource(cluster, groupResource)
	if err != nil {
		return err
	}

	for _, binding := range bindings {
		for _, claim := range binding.Spec.PermissionClaims {
			if claim.State != apisv1alpha1.ClaimAccepted || claim.Group != groupResource.Group || claim.Resource != groupResource.Resource {
				// There is no PermissionClaim for this object, no need to compute selection
				continue
			}

			visit(binding, claim, resourceNamespace, resourceName)
		}
	}
}

// LabelsFor returns all the applicable labels for the cluster-group-resource relating to permission claims. This is
// the intersection of (1) all APIBindings in the cluster that have accepted claims for the group-resource with (2)
// associated APIExports that are claiming group-resource and possibly namespace/name.
func (l *Labeler) LabelsFor(ctx context.Context, cluster logicalcluster.Name, groupResource schema.GroupResource, resourceName, resourceNamespace string) (map[string]string, error) {
	labels := map[string]string{}

	logger := klog.FromContext(ctx)

	l.visitBindingsAndClaims(ctx, cluster, groupResource, resourceNamespace, resourceName, func(binding *apisv1alpha1.APIBinding,
		claim apisv1alpha1.AcceptablePermissionClaim, namespace,
		name string) {
		path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path)
		if path.Empty() {
			path = logicalcluster.From(binding).Path()
		}
		export, err := l.getAPIExport(path, binding.Spec.Reference.Export.Name)
		if err != nil {
			return
		}
		// The permission claim is relevant for this object, but the object does not match the criteria
		if !Selects(claim, resourceName, resourceNamespace) {
			return
		}
		k, v, err := permissionclaims.ToLabelKeyAndValue(logicalcluster.From(export), export.Name, claim.PermissionClaim)
		if err != nil {
			// extremely unlikely to get an error here - it means the json marshaling failed
			logger.Error(err, "error calculating permission claim label key and value",
				"claim", claim.String())
			return
		}
		labels[k] = v
	})

	// for APIBindings we have to set the constant label value on itself to make the APIBinding
	// pointing to an APIExport visible to the owner of the export, independently of the permission claim
	// acceptance of the binding.
	if groupResource.Group == apis.GroupName && groupResource.Resource == "apibindings" {
		binding, err := l.getAPIBinding(cluster, resourceName)
		if err != nil {
			logger.Error(err, "error getting APIBinding", "bindingName", resourceName)
			return labels, nil // can only be a NotFound
		}

		path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path)
		if path.Empty() {
			path = logicalcluster.From(binding).Path()
		}
		export, err := l.getAPIExport(path, binding.Spec.Reference.Export.Name)
		if err == nil {
			k, v := permissionclaims.ToReflexiveAPIBindingLabelKeyAndValue(logicalcluster.From(export), binding.Spec.Reference.Export.Name)
			if _, found := labels[k]; !found {
				labels[k] = v
			}
		}
	}

	return labels, nil
}

// ObjectPermitted determines if any PermissionClaims apply to a given object's GroupResource, namespace, or name.
// If no APIBinding or PermissionClaim exists for the GroupResource, the object is permitted.
func (l *Labeler) ObjectPermitted(ctx context.Context, cluster logicalcluster.Name, groupResource schema.GroupResource, resourceName, resourceNamespace string) (bool, error) {
	permitted := true
	l.visitBindingsAndClaims(ctx, cluster, groupResource, resourceNamespace, resourceName, func(apiBinding *apisv1alpha1.APIBinding, claim apisv1alpha1.AcceptablePermissionClaim, namespace, name string) {
		if !Selects(claim, resourceName, resourceNamespace) {
			// The permission claim is relevant for this object, but the object does not match the criteria
			permitted = false
		}
	})
	return permitted, nil
}
```